worlds/zork_grand_inquisitor/game_controller.py

```diff
@@ -0,0 +1,1447 @@
+                if ZorkGrandInquisitorLocations.YOU_GAINED_86_EXPERIENCE_POINTS in self.completed_locations:
+                    if self._player_has(ZorkGrandInquisitorItems.SPELL_GLORF):
+                        self.completed_locations_queue.append(missable_location)
+            elif missable_location == ZorkGrandInquisitorLocations.WHAT_ARE_YOU_STUPID:
+                if ZorkGrandInquisitorLocations.FIRE_FIRE in self.completed_locations:
+                    if self._player_has(ZorkGrandInquisitorItems.PLASTIC_SIX_PACK_HOLDER):
+                        self.completed_locations_queue.append(missable_location)
+            elif missable_location == ZorkGrandInquisitorLocations.YAD_GOHDNUORGREDNU_3_YRAUBORF:
+                if ZorkGrandInquisitorLocations.REASSEMBLE_SNAVIG in self.completed_locations:
+                    self.completed_locations_queue.append(missable_location)
+            elif missable_location == ZorkGrandInquisitorLocations.YOUR_PUNY_WEAPONS_DONT_PHASE_ME_BABY:
+                if ZorkGrandInquisitorLocations.WANT_SOME_RYE_COURSE_YA_DO in self.completed_locations:
+                    if self._player_has(ZorkGrandInquisitorItems.SWORD) and self._player_has(
+                            ZorkGrandInquisitorItems.HOTSPOT_HARRY
+                    ):
+                        self.completed_locations_queue.append(missable_location)
+            elif missable_location == ZorkGrandInquisitorLocations.YOU_DONT_GO_MESSING_WITH_A_MANS_ZIPPER:
+                if ZorkGrandInquisitorLocations.YOU_GAINED_86_EXPERIENCE_POINTS in self.completed_locations:
+                    if self._player_has(ZorkGrandInquisitorItems.SPELL_REZROV):
+                        self.completed_locations_queue.append(missable_location)
+            elif missable_location == ZorkGrandInquisitorLocations.YOU_WANT_A_PIECE_OF_ME_DOCK_BOY:
+                if ZorkGrandInquisitorLocations.HELP_ME_CANT_BREATHE in self.completed_locations:
+                    self.completed_locations_queue.append(missable_location)
+
+    def _process_received_items(self) -> None:
+        while len(self.received_items_queue) > 0:
+            item: ZorkGrandInquisitorItems = self.received_items_queue.popleft()
+            data: ZorkGrandInquisitorItemData = item_data[item]
+
+            if ZorkGrandInquisitorTags.FILLER in data.tags:
+                continue
+
+            self.received_items.add(item)
+
+    def _manage_hotspots(self) -> None:
+        hotspot_item: ZorkGrandInquisitorItems
+        for hotspot_item in self.all_hotspot_items:
+            data: ZorkGrandInquisitorItemData = item_data[hotspot_item]
+
+            if hotspot_item not in self.received_items:
+                key: int
+                for key in data.statemap_keys:
+                    self._write_game_flags_value_for(key, 2)
+            else:
+                if hotspot_item == ZorkGrandInquisitorItems.HOTSPOT_666_MAILBOX:
```

Comment:
kind of the same comment here. i think a lot of this can be reduced down and made more readable with dicts or maybe a dataclass.

Reply:
I disagree here. The huge if trees are absolutely ugly, no question there, but they are certainly readable (from my perspective as the person that will have to maintain this). Readability isn't just how concise the code it, it's also how quickly you can grasp what the code is meant to do. Would it be possible to abstract it to a single data structure, I would do it (like I did for the grant conditions, thanks to your suggestion), but simplifying only the parts that can be simplified implies that the logic will be split between 2 places and that actively harms readability

Reply:
i mean i did and it's definitely doable, but i do agree that it's likely not worth the effort and even then it wouldn't necessarily guarantee more readability due to all the specific conditions